transform/transform.go

```diff
@@ -19,26 +19,16 @@ package transform
 
 import (
 	"regexp"
-	"time"
 
-	"github.com/elastic/beats/libbeat/beat"
-
-	"github.com/elastic/apm-server/model/metadata"
 	"github.com/elastic/apm-server/sourcemap"
+	"github.com/elastic/beats/libbeat/beat"
 )
 
 type Transformable interface {
-	Transform(*Context) []beat.Event
-}
-
-type Context struct {
-	RequestTime time.Time
-	Config      Config
-	Metadata    metadata.Metadata
+	Transform(Config, *sourcemap.Store) []beat.Event
```

Comment:
:100: on removing `transform.Context`, but I think we should go further and remove the remaining config and sourcemap params. Reason being that these do not differ per event.

How about this as an alternative?

If we follow with through the `Decoder` interface idea, then we can have concrete types like `StacktraceDecoder`. That type could have a field for storing a `sourcemap.Store`, and the `Stacktrace` objects returned by `Decode` would keep an internal reference to it (either directly or by referencing the `StacktraceDecoder`). i.e. something like this:

```go
type StacktraceDecoder struct {
    SourcemapStore *sourcemap.Store
    LibraryPattern *regexp.Regexp
    ExcludeFromGrouping *regexp.Regexp
}

type Stacktrace struct {
    decoder *StacktraceDecoder
    Frames []*StacktraceFrame
}

func (s *Stacktrace) Transform(context.Context) []beat.Event {
    // Apply source maps using s.decoder.SourcemapStore.
}
```

Additional `ErrorDecoder` and `SpanDecoder` types would each have `StacktraceDecoder`-type fields. Rather than constructing all of these things globally, they would be constructed in `beater/api/mux.go`, passing in config at construction time.

Note I added a context.Context param to the Transform method. This is not necessary for this PR, but it will be needed for https://github.com/elastic/apm-server/issues/3184 and if we're changing them all then it might be a good opportunity to add it in.

Reply:
It feels a bit creepy to me to have the sourcemap thing (basically global mutable state) as an attribute hidden down in some data type, i rather pass it explicitly as function argument top level.
That said I don't like neither to pass arguments that are not needed in most implementations... I'll explore the `Decoder` idea.